use_cases/eluc/app/utils.py

```diff
@@ -1,21 +1,116 @@
+
 import pandas as pd
+import torch
+
+from prsdk.persistence.serializers.neural_network_serializer import NeuralNetSerializer
+from prsdk.persistence.persistors.hf_persistor import HuggingFacePersistor
 
-import app.constants as app_constants
 from data import constants
+from data.eluc_encoder import ELUCEncoder
+from prescriptors.prescriptor_manager import PrescriptorManager
+from prescriptors.nsga2.candidate import Candidate
+from prescriptors.nsga2.land_use_prescriptor import LandUsePrescriptor
+from predictors.percent_change.percent_change_predictor import PercentChangePredictor
+
+
+def add_nonland(context_actions_df: pd.DataFrame) -> pd.DataFrame:
+    """
+    Adds a nonland column to the context_actions_df
+    """
+    updated_df = context_actions_df.copy()
+    updated_df["nonland"] = 1 - updated_df[constants.LAND_USE_COLS].sum(axis=1)
+    updated_df["nonland"] = updated_df["nonland"].clip(lower=0)
+    return updated_df
 
 
-def add_nonland(data: pd.Series) -> pd.Series:
+class EvolutionHandler:
     """
-    Adds a nonland column that is the difference between 1 and
-    LAND_USE_COLS.
-    Note: Since sum isn't exactly 1 we just set to 0 if we get a negative.
-    :param data: pd Series containing land use data.
-    :return: pd Series with nonland column added.
+    Class that handles running our trained prescriptors
     """
-    data = data[constants.LAND_USE_COLS]
-    nonland = 1 - data.sum() if data.sum() <= 1 else 0
-    data['nonland'] = nonland
-    return data[app_constants.CHART_COLS]
+
+    def __init__(self):
+        self.prescriptor_path = Path("app/results/prescriptors")
+        self.prescriptor_list = [model_path.stem for model_path in self.prescriptor_path.glob("*.pt")]
+
+        def sort_cand_id(cand_id: str):
+            return (int(cand_id.split("_")[0]), int(cand_id.split("_")[1]))
+        self.prescriptor_list = sorted(self.prescriptor_list, key=sort_cand_id)
+
+        # Load prescriptors
+        self.prescriptor_manager = self.load_prescriptors()
+
+    def load_prescriptors(self) -> PrescriptorManager:
+        """
+        Loads in prescriptors from a hard-coded location on disk.
+        Loads predictor from HuggingFace if needed.
+        :return: dict of prescriptor name -> prescriptor object.
+        """
+        with open("app/results/fields.json", "r", encoding="utf-8") as f:
+            fields = json.load(f)
+        encoder = ELUCEncoder(fields)
+
+        candidate_params = {"in_size": len(constants.CAO_MAPPING["context"]),
+                            "hidden_size": 16,
+                            "out_size": len(constants.RECO_COLS)}
+        prescriptors = {}
+        for cand_id in self.prescriptor_list:
+            cand_path = self.prescriptor_path / f"{cand_id}.pt"
+            candidate = Candidate(**candidate_params, cand_id=cand_id)
+            candidate.load_state_dict(torch.load(cand_path))
+            prescriptors[cand_id] = LandUsePrescriptor(candidate, encoder)
```

Comment:
For now the prescriptors are hard-coded and saved in the repo. In the future these should live in HuggingFace